Make it possible to get the scrolling tree with UI-side compositing on macOS

smfr opened the pull request:
#### e4fa5c5f61bb9b6d345275cf05d8246fb5ac9952
<pre>
Make it possible to get the scrolling tree with UI-side compositing on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=251193">https://bugs.webkit.org/show_bug.cgi?id=251193</a>
rdar://104681297

Reviewed by Tim Horton.

`window.internals.scrollingTreeAsText()` only works when we have a scrolling tree in the web process,
i.e. only with non-UI-side compositing. In order to make tests work with both UI-side and non-UI-side
compositing, add `internals.haveScrollingTree()` and use it in `UIHelper.getScrollingTree()` to know
when we should use UIScriptController to fetch the scrolling tree from the UI process.

Make the existing WKWebView testing code for getting the scrolling tree work on both
macOS and iOS.

Fix up one test (fixed-background-in-overflow-in-iframe.html) to use the new thing; other
tests will be fixed soon.

* LayoutTests/platform/mac-gpup/scrollingcoordinator/mac/fixed-backgrounds/fixed-background-in-overflow-in-iframe-expected.txt: Copied from LayoutTests/scrollingcoordinator/mac/fixed-backgrounds/fixed-background-in-overflow-in-iframe-expected.txt.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.getScrollingTree):
* LayoutTests/scrollingcoordinator/mac/fixed-backgrounds/fixed-background-in-overflow-in-iframe-expected.txt:
* LayoutTests/scrollingcoordinator/mac/fixed-backgrounds/fixed-background-in-overflow-in-iframe.html:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::haveScrollingTree const):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::haveScrollingTree const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::haveScrollingTree const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _scrollingTreeAsText]):
* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _scrollingTreeAsText]): Deleted.
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::scrollingTreeAsText const):
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.h:

Canonical link: <a href="https://commits.webkit.org/259457@main">https://commits.webkit.org/259457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca5b9a5a75f01b4c1980b17f9c8dd9a0252e87b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104903 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13983 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37798 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114170 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174360 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108808 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15115 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4907 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97229 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113192 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110663 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11679 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94686 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39198 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93549 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26307 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80860 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7327 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27666 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7420 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4254 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13478 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47218 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6527 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9212 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:
```diff
@@ -321,6 +321,11 @@ static void playBackEvents(WKWebView *webView, UIScriptContext *context, NSStrin
 
+JSRetainPtr<JSStringRef> UIScriptControllerMac::scrollingTreeAsText() const
+{
+    return adopt(JSStringCreateWithCFString((CFStringRef)[webView() _scrollingTreeAsText]));
```
why is this not in some Cocoa file??

smfr replied:
Will fix.

smfr committed:
Make it possible to get the scrolling tree with UI-side compositing on macOS
https://bugs.webkit.org/show_bug.cgi?id=251193
rdar://104681297

Reviewed by Tim Horton.

`window.internals.scrollingTreeAsText()` only works when we have a scrolling tree in the web process,
i.e. only with non-UI-side compositing. In order to make tests work with both UI-side and non-UI-side
compositing, add `internals.haveScrollingTree()` and use it in `UIHelper.getScrollingTree()` to know
when we should use UIScriptController to fetch the scrolling tree from the UI process.

Make the existing WKWebView testing code for getting the scrolling tree work on both
macOS and iOS.

Fix up one test (fixed-background-in-overflow-in-iframe.html) to use the new thing; other
tests will be fixed soon.

* LayoutTests/platform/mac-gpup/scrollingcoordinator/mac/fixed-backgrounds/fixed-background-in-overflow-in-iframe-expected.txt: Copied from LayoutTests/scrollingcoordinator/mac/fixed-backgrounds/fixed-background-in-overflow-in-iframe-expected.txt.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.getScrollingTree):
* LayoutTests/scrollingcoordinator/mac/fixed-backgrounds/fixed-background-in-overflow-in-iframe-expected.txt:
* LayoutTests/scrollingcoordinator/mac/fixed-backgrounds/fixed-background-in-overflow-in-iframe.html:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::haveScrollingTree const):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::haveScrollingTree const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::haveScrollingTree const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _scrollingTreeAsText]):
* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _scrollingTreeAsText]): Deleted.
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::scrollingTreeAsText const):
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.h:

Canonical link: https://commits.webkit.org/259457@main
